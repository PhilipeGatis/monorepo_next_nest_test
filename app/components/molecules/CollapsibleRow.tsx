import React, { useState } from 'react'
import TableRow from '@mui/material/TableRow'
import TableCell from '@mui/material/TableCell'
import IconButton from '@mui/material/IconButton'
import KeyboardArrowUpIcon from '@mui/icons-material/KeyboardArrowUp'
import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown'
import Collapse from '@mui/material/Collapse'
import Box from '@mui/material/Box'
import Table from '@mui/material/Table'
import TableHead from '@mui/material/TableHead'
import TableBody from '@mui/material/TableBody'
import { parseJSON, format } from 'date-fns'
import { ProductListGroupedByDescriptionResponseDto } from '@hubla/api/dist/models/product/dto/ProductListGroupedByDescriptionResponse.dto'
import { ProductResponseDto } from '@hubla/api/dist/models/product/dto/ProductResponse.dto'

const calcTotal = (list: ProductResponseDto[]): string => {
  let sum = 0
  for (const product of list) {
    if (product.type.signal) {
      sum = sum + product.value
    } else {
      sum = sum - product.value
    }
  }
  return sum.toLocaleString('pt-BR', {
    style: 'currency',
    minimumFractionDigits: 2,
    maximumFractionDigits: 3,
    currency: 'BRL',
  })
}

const CollapsibleRow: React.FC<{ row: ProductListGroupedByDescriptionResponseDto }> = ({ row }) => {
  const [open, setOpen] = useState(false)

  return (
    <>
      <TableRow sx={{ '& > *': { borderBottom: 'unset' } }}>
        <TableCell>
          <IconButton aria-label="expand row" size="small" onClick={() => setOpen(!open)}>
            {open ? <KeyboardArrowUpIcon /> : <KeyboardArrowDownIcon />}
          </IconButton>
        </TableCell>
        <TableCell component="th" scope="row">
          {row.title}
        </TableCell>
        <TableCell align="right">{calcTotal(row.list)}</TableCell>
      </TableRow>
      <TableRow>
        <TableCell style={{ paddingBottom: 0, paddingTop: 0 }} colSpan={6}>
          <Collapse in={open} timeout="auto" unmountOnExit>
            <Box sx={{ margin: 1 }}>
              <Table size="small" aria-label="purchases">
                <TableHead>
                  <TableRow>
                    <TableCell>Vendedor</TableCell>
                    <TableCell>Descrição</TableCell>
                    <TableCell>Natureza</TableCell>
                    <TableCell>data</TableCell>
                    <TableCell align="right">Valor</TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {row.list.map(product => {
                    const isPositive = product.type.signal
                    return (
                      <TableRow key={product.id}>
                        <TableCell component="th" scope="row" sx={{ color: isPositive ? 'green' : 'red' }}>
                          {product.seller}
                        </TableCell>
                        <TableCell component="th" scope="row" sx={{ color: isPositive ? 'green' : 'red' }}>
                          {product.type.descriptions}
                        </TableCell>
                        <TableCell component="th" scope="row" sx={{ color: isPositive ? 'green' : 'red' }}>
                          {product.type.nature}
                        </TableCell>
                        <TableCell sx={{ color: isPositive ? 'green' : 'red' }}>
                          {format(parseJSON(product.date), 'Pp')}
                        </TableCell>
                        <TableCell align="right" sx={{ color: isPositive ? 'green' : 'red' }}>
                          {(product.value * (isPositive ? 1 : -1)).toLocaleString('pt-BR', {
                            style: 'currency',
                            minimumFractionDigits: 2,
                            maximumFractionDigits: 3,
                            currency: 'BRL',
                          })}
                        </TableCell>
                      </TableRow>
                    )
                  })}
                </TableBody>
              </Table>
            </Box>
          </Collapse>
        </TableCell>
      </TableRow>
    </>
  )
}

export default CollapsibleRow
