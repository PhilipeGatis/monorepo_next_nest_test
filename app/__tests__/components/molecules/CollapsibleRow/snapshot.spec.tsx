import { render } from '@testing-library/react'
import CollapsibleRow from '../../../../components/molecules/CollapsibleRow'
import mockData from './mock'

describe('CollapsibleRow', () => {
  it('renders component', () => {
    const { container } = render(<CollapsibleRow row={mockData} />)
    expect(container).toMatchSnapshot()
  })
})
