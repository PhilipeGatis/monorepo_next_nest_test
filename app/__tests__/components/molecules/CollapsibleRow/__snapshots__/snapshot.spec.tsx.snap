// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapsibleRow renders component 1`] = `
<div>
  <tr
    class="MuiTableRow-root css-2blr9s-MuiTableRow-root"
  >
    <td
      class="MuiTableCell-root MuiTableCell-sizeMedium css-bsph2u-MuiTableCell-root"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
        data-testid="expandButton"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="KeyboardArrowDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </td>
    <th
      class="MuiTableCell-root MuiTableCell-sizeMedium css-bsph2u-MuiTableCell-root"
      scope="row"
    >
      CURSO DE BEM-ESTAR
    </th>
    <td
      class="MuiTableCell-root MuiTableCell-alignRight MuiTableCell-sizeMedium css-r7oxtk-MuiTableCell-root"
    >
      R$Â 25.000,00
    </td>
  </tr>
  <tr
    class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
  >
    <td
      class="MuiTableCell-root MuiTableCell-sizeMedium css-bsph2u-MuiTableCell-root"
      colspan="6"
      style="padding-bottom: 0px; padding-top: 0px;"
    />
  </tr>
</div>
`;
